cmds/dd/dd.go

```diff
@@ -25,56 +25,84 @@ package main
 import (
 	"bytes"
 	"flag"
-	"fmt"
 	"io"
 	"log"
 	"math"
 	"os"
 	"strings"
 )
 
+type data struct {
+	sz   int64
+	data []byte
+}
+
 var (
 	ibs     = flag.Int64("ibs", 1, "Default input block size")
 	obs     = flag.Int64("obs", 1, "Default output block size")
 	bs      = flag.Int64("bs", 0, "Default input and output block size")
 	skip    = flag.Int64("skip", 0, "skip n bytes before reading")
 	seek    = flag.Int64("seek", 0, "seek output when writing")
-	conv    = flag.String("conv", "", "Convert the file on a specific way, like notrunc")
+	conv    = flag.String("conv", "none", "Convert the file on a specific way, like notrunc")
 	count   = flag.Int64("count", math.MaxInt64, "Max output of data to copy")
 	inName  = flag.String("if", "", "Input file")
 	outName = flag.String("of", "", "Output file")
+	convs   = map[string]func([]byte) []byte{
+		"none":  func(b []byte) []byte { return b },
+		"ucase": bytes.ToUpper,
+		"lcase": bytes.ToLower,
+	}
+	convert func([]byte) []byte
 )
 
 func pass(r io.Reader, w io.Writer, ibs, obs int64, conv string) {
-	var b = &bytes.Buffer{}
-	for {
-		b.Reset()
-		n, err := io.CopyN(b, r, ibs)
-		if n == 0 {
-			break
-		}
-		if err != nil && err != io.EOF {
-			log.Fatalf("Reading from input: %v", err)
-		}
-		var out *bytes.Buffer
-		switch conv {
-		case "ucase":
-			out = bytes.NewBuffer(bytes.ToUpper(b.Bytes()))
-		case "lcase":
-			out = bytes.NewBuffer(bytes.ToLower(b.Bytes()))
-		default:
-			out = bytes.NewBuffer(b.Bytes())
-		}
-		for n := int64(0); n < int64(out.Len()); {
-			amt, err := io.CopyN(w, out, obs)
+	// make the channels deep enough for 1 GiB
+	depth := (1024 * 1048576) / ibs 
+	// but keep it reasonable!
+	if depth > 8192 {
+		depth = 8192
+	}
+	bufs := make(chan *data, depth)
+	i := make(chan *data, depth)
+	go func() {
+		var d *data
+		for {
+			select {
+			case d = <-bufs:
+			default:
+				d = &data{ibs, make([]byte, ibs, ibs)}
```

Comment:
I suspect between the concurrency overhead + stress on the garbage collector, this implementation is a lot slower than coreutils dd.